orc8r/cloud/go/syncstore/store_writer.go

```diff
@@ -0,0 +1,393 @@
+	if err != nil {
+		return errors.Wrap(err, "error starting transaction")
+	}
+	defer store.Rollback()
+
+	inStore, err := getInStoreNetworksBlobstore(store)
+	if err != nil {
+		return errors.Wrap(err, "get all networks in blobstore")
+	}
+	deleted, _ := funk.DifferenceString(inStore, tracked)
+
+	errs := &multierror.Error{}
+	for _, network := range deleted {
+		keys, err := blobstore.ListKeys(store, network, lastResyncBlobstoreType)
+		if err != nil {
+			multierror.Append(errs, err)
+			continue
+		}
+		err = store.Delete(network, storage.MakeTKs(lastResyncBlobstoreType, keys))
+		if err != nil {
+			multierror.Append(errs, err)
+		}
+	}
+	err = store.Commit()
+	if err != nil {
+		multierror.Append(errs, err)
+	}
+	return errs.ErrorOrNil()
+}
+
+func (l *syncStore) collectGarbageCacheWriter(tracked []string) error {
+	errs := &multierror.Error{}
+
+	invalidByNetwork, err := getInvalidCacheWriter(l.fact, tracked, l.cacheWriterValidIntervalSecs)
+	if err != nil {
+		multierror.Append(errs, errors.Wrapf(err, "get invalid cache writers for tracked networks %+v", tracked))
+	}
+
+	// Attempt to drop the tmp tables of all invalid cacheWriters, and only delete the blobstore records of those
+	// whose tables have been successfully dropped; the rest is left to be garbage collected in future runs
+	deletedByNetwork, err := dropInvalidCaches(l.db, invalidByNetwork)
+	if err != nil {
+		multierror.Append(errs, errors.Wrapf(err, "drop invalid cache writer tables %+v", invalidByNetwork))
+	}
+
+	err = deleteCacheWriterBlobstoreRecords(l.fact, deletedByNetwork)
+	if err != nil {
+		multierror.Append(errs, errors.Wrapf(err, "delete cache writer blobstore records %+v", deletedByNetwork))
+	}
+
+	return errs.ErrorOrNil()
+}
+
+// getInvalidCacheWriter returns a list of cache writer IDs from blobstore that
+// either belong to already deleted networks or have expired.
+func getInvalidCacheWriter(fact blobstore.BlobStorageFactory, tracked []string, cacheWriterValidIntervalSecs int64) (map[string][]string, error) {
+	store, err := fact.StartTransaction(nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "error starting transaction")
+	}
+	defer store.Rollback()
+
+	inStore, err := getInStoreNetworksBlobstore(store)
+	if err != nil {
+		return nil, errors.Wrap(err, "get all networks in blobstore")
+	}
+
+	deleted, _ := funk.DifferenceString(inStore, tracked)
+
+	invalidByNetwork := map[string][]string{}
+	errs := &multierror.Error{}
+	for _, network := range deleted {
+		keys, err := blobstore.ListKeys(store, network, cacheWriterBlobstoreType)
+		if err != nil {
+			multierror.Append(errs, errors.Wrapf(err, "list cache writers of deleted network %+v", network))
+			continue
+		}
+		invalidByNetwork[network] = keys
+	}
+
+	for _, network := range tracked {
+		filter := blobstore.CreateSearchFilter(&network, []string{cacheWriterBlobstoreType}, nil, nil)
+		criteria := blobstore.LoadCriteria{LoadValue: true}
+		networkBlobs, err := store.Search(filter, criteria)
```

Comment:
There's probably a static method you can use that wraps Search, to make this cleaner